DIPs/1NNN-DK.md

```diff
@@ -0,0 +1,242 @@
+
+```D
+struct S
+{
+    @nogc nothrow pure @safe:
+    private:
+    const:
+
+    void f()
+    {
+
+    }
+
+    void g()
+    {
+
+    }
+
+    // ...
+}
+// attributes' influence ends here
+```
+
+This reduces the repetition of attributes, but it becomes a problem when there are a few declarations that need an exception.
+Sometimes this can be solved by annotating the exceptional declaration with an inverse attribute:
+
+| Attribute                | Undone by           |
+|--------------------------|---------------------|
+| `private`                | `public`            |
+| `@system`                | `@safe`             |
+| `extern(C)`              | `extern(D)`         |
+| `nothrow`                | `throw`             |
+| `pragma(inline, true)`   | `pragma(inline)`    |
+
+However, most attributes do not have an inverse that resets it to the default:
+
+- `align()`
+- `deprecated`
+- `static`
+- `extern`
+- `abstract`
+- `final`
+- `override`
+- `synchronized`
+- `scope`
+- `const`
+- `immutable`
+- `inout`
+- `shared`
+- `__gshared`
+- `pure`
+- `@disable`
+- `@nogc`
+- `@property`
+- `@UserDefinedAttribute`
+
+Moreover, because function attributes are inferred in template functions, forcing them back to the default is not always desirable either.
+Consider:
+
+```
+struct S
+{
+    @nogc nothrow pure @safe:
+
+    void f() { }
+
+    void g() { }
+
+    void toString(O)(ref O sink) if (isOutputRange!(O, char))
+    {
+    }
+}
+```
+
+In this case, the `toString` method should not force the provided Output Range to have an implementation of `sink.put` with all function attributes, but
+neither should `toString` be denied those attributes when it can infer them.
+The only option here is to place all template methods above the `@nogc nothrow pure @safe:` line, or go back to 'attribute soup' by repeating all function attributes for individual members.
+
+What is really desired is something to specify that attributes on `toString` should go back to the default state,
+which this DIP proposes in the form of the `@default` attribute.
+
+## Prior Work
+
+- [DIP 1029 - Add throw as Function Attribute](http://dlang.org/dips/1029)
+The `throw` attribute was added to address the lack of a way to invert the effect of `nothrow:`.
+However, it has not been implemented yet, and can be superseded by this DIP.
+
+- [DIP 1012 - Function Attributes](http://dlang.org/dips/1012)
+Proposes to move function attributes to `core.attribute` enabling traditional methods of attribute manipulation (`AliasSeq`) and introspection.
+This DIP has been abandoned.
+
+## Alternatives
+
+It has been proposed before that attributes could take a boolean parameter evaluated at CTFE (e.g. `@safe(x && y)`), or reset to the default by passing the `default` keyword as a parameter.
+This allows more precision, but it is more complex and verbose:
+
+```D
+@nogc nothrow pure @safe:
+
+// ...
+
+void f(T)(T x) @nogc(default) @nothrow(default) pure(default) @safe(default)
```

Comment:
On a template, it is not clear what `@safe(default)` even means. On a regular function, `@safe(default)` is the same as `@system`, but on a template, it seems to be mean infer. Better: Just use “infer” instead of “default” everywhere or make “default” mean “infer attributes” everywhere.